staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1338,3 +1477,110 @@ func buildOpenAPIModelsForApply(staticOpenAPISpec *goopenapispec.Swagger, crd *a
 	}
 	return models, nil
 }
+
+func (r *crdHandler) updateStorageVersionFor(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	gr := schema.GroupResource{
+		Group:    crd.Spec.Group,
+		Resource: crd.Spec.Names.Plural,
+	}
+	storageVersion, err := apiextensionshelpers.GetCRDStorageVersion(crd)
+	if err != nil {
+		// This should never happened if crd is valid, which is true since we
+		// only update storage version for CRDs that have been written to the
+		// storage
+		return err
+	}
+	encodingVersion := crd.Spec.Group + "/" + storageVersion
+	var decodableVersions []string
+	for _, v := range crd.Spec.Versions {
+		decodableVersions = append(decodableVersions, crd.Spec.Group+"/"+v.Name)
+	}
+
+	appendOwnerRefFunc := func(sv *apiserverinternalv1alpha1.StorageVersion) error {
+		ref := metav1.OwnerReference{
+			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+			Name:       crd.Name,
+			UID:        crd.UID,
+		}
+		for _, r := range sv.OwnerReferences {
+			if r == ref {
+				return nil
+			}
+		}
+		sv.OwnerReferences = append(sv.OwnerReferences, ref)
+		return nil
+	}
+	return storageversion.UpdateStorageVersionFor(
+		ctx,
+		r.storageVersionClient,
+		r.apiserverID,
+		gr,
+		encodingVersion,
+		decodableVersions,
+		appendOwnerRefFunc)
+}
+
+// getOrCreateStorageVersionUpdateQueue_locked returns the channel for the given UID,
+// or create a new one and a new goroutine if necessary. The goroutine keeps processing
+// requests until the channel is closed.
+// NOTE: caller should hold the customStorageLock until the write to the returned channel is finished,
+// to make sure the storage version update is in sync with the underlying storage change.
+func (r *crdHandler) getOrCreateStorageVersionUpdateQueue_locked(uid types.UID) chan *storageVersionUpdateRequest {
+	if queue, ok := r.storageVersionUpdateQueues[uid]; ok {
+		return queue.q
+	}
+
+	queue := make(chan *storageVersionUpdateRequest, maxStorageVersionUpdateRequests)
+	ctx, cancel := context.WithCancel(context.TODO())
+	r.storageVersionUpdateQueues[uid] = &storageVersionUpdateQueue{
+		q:      queue,
+		cancel: cancel,
+	}
+	go func() {
+		for req := range queue {
+			if req.wait != nil {
+				<-req.wait
+			}
+			if err := r.updateStorageVersionFor(ctx, req.crd); err != nil {
+				utilruntime.HandleError(err)
+			}
+			if req.processed != nil {
+				select {
+				case <-ctx.Done():
```

Comment:
Why not checking the ctx.Done() at the beginning? Then we can skip updating storage version if the CRD is being deleted.

Reply:
> skip updating storage version if the CRD is being deleted

client-go is responsible for doing that. 

If we check ctx.Done() at the beginning, there could be a case where the CRD is being deleted after the check but before `r.updateStorageVersionFor`. Still, client-go has to evaluate the context and skip updating storage version.

Reply:
Got it. Can you test if client-go actually respects this ctx?

Reply:
Yes, it does
```
--- a/staging/src/k8s.io/client-go/examples/out-of-cluster-client-configuration/main.go
+++ b/staging/src/k8s.io/client-go/examples/out-of-cluster-client-configuration/main.go
@@ -61,7 +61,9 @@ func main() {
                panic(err.Error())
        }
        for {
-               pods, err := clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+               ctx, cancel := context.WithCancel(context.Background())
+               cancel()
+               pods, err := clientset.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
                if err != nil {
                        panic(err.Error())
                }

$ go run examples/out-of-cluster-client-configuration/main.go --kubeconfig=/var/run/kubernetes/admin.kubeconfig
panic: context canceled
```

Reply:
Thanks for the confirmation.

Adding a check of ctx.Done() at the beginning let us skip unnecessary wait on `<-req.wait`.